runtime-sdk/src/modules/accounts/mod.rs

```diff
@@ -481,18 +561,102 @@ impl module::AuthHandler for Module {
         let mut store = storage::PrefixStore::new(ctx.runtime_state(), &MODULE_NAME);
         let mut accounts =
             storage::TypedStore::new(storage::PrefixStore::new(&mut store, &state::ACCOUNTS));
+        let mut payee = None;
         for si in tx.auth_info.signer_info.iter() {
             let address = Address::from_pk(&si.public_key);
             let mut account: types::Account = accounts.get(&address).unwrap_or_default();
             if account.nonce != si.nonce {
                 return Err(modules::core::Error::InvalidNonce);
             }
 
+            // First signer pays for the fees.
+            if payee.is_none() {
+                payee = Some(address);
+            }
+
             // Update nonce.
             // TODO: Could support an option to defer this.
             account.nonce += 1;
             accounts.insert(&address, &account);
         }
+
+        // Charge the specified amount of fees.
+        if !tx.auth_info.fee.amount.amount().is_zero() {
+            let payee = payee.expect("at least one signer is always present");
+
+            Self::sub_amount(ctx.runtime_state(), payee, &tx.auth_info.fee.amount)
+                .map_err(|_| modules::core::Error::InsufficientFeeBalance)?;
+
+            ctx.value::<FeeAccumulator>(CONTEXT_KEY_FEE_ACCUMULATOR)
+                .add(&tx.auth_info.fee.amount);
+
+            // TODO: Emit event that fee has been paid.
+        }
         Ok(())
     }
 }
+
+impl module::BlockHandler for Module {
+    fn end_block(ctx: &mut DispatchContext) {
+        // Determine the fees that are available for disbursement from the last block.
+        let mut previous_fees = Self::get_balances(ctx.runtime_state(), *ADDRESS_FEE_ACCUMULATOR)
+            .expect("get_balances must succeed")
+            .balances;
+
+        // Disburse transaction fees to entities controlling all the good nodes in the committee.
+        let addrs: Vec<Address> = ctx
+            .runtime_round_results()
+            .good_compute_entities
+            .iter()
+            .map(|pk| Address::from_pk(&PublicKey::Ed25519(pk.into())))
+            .collect();
+
+        if !addrs.is_empty() {
+            let amounts: Vec<_> = previous_fees
+                .iter()
+                .map(|(denom, fee)| {
+                    token::BaseUnits::new(
+                        fee.checked_div(&token::Quantity::from(addrs.len() as u64))
+                            .expect("addrs is non-empty"),
+                        denom.clone(),
+                    )
+                })
```

Comment:
Could we also filter out zero amounts here, to avoid unnecessary adds/subs bellow?

Reply:
Good idea, will change it.

Reply:
Done.